Initial scale of a page is not respected when both initial scale and viewport width are set

aprotyas opened the pull request:
#### 524f1bcc6830730eb57dc4458b500f1f4f051e2e
<pre>
Initial scale of a page is not respected when both initial scale and viewport width are set
<a href="https://bugs.webkit.org/show_bug.cgi?id=258663">https://bugs.webkit.org/show_bug.cgi?id=258663</a>
rdar://111515914

Reviewed by Wenson Hsieh.

In 256476@main, we introduce a viewport sizing heuristic to clamp the
maximum initial scale of a page. This was necessitated by Wikipedia&apos;s
dynamic initial scale assignment (to maintain their viewport width
invariant of 1000) leading to Wiki articles being excessively zoomed in
on large screens (such as when an iPad is connected to an external
display).

However, the somewhat arbitrarily chosen initial scale cap of 1.2 leads
to a sub-optimal web development experience because we perform the
initial scale clamping on every page (when the `IgnoreMetaViewport` is
enabled), meaning it isn&apos;t clear why the initial scale is not being
respected, even when a developer may have had a valid use-case.

To mitigate this situation, we only apply this initial scale clamping as
a quirk for pages on the &quot;wikipedia.org&quot; domain. This is a relatively
safe decision given that Wikipedia is probably the only top site that
performs a dynamic initial scale assignment (that too with potentially
*large* scale values).

We also introduce an API test to assert this behavior is only limited to
Wikipedia.

* LayoutTests/fast/viewport/ios/adjust-fixed-width-and-initial-scale-to-avoid-excessive-zooming.html: Removed.

This layout test is no longer valid.

* Source/WebCore/page/Quirks.cpp:
(WebCore::isWikipediaDomain):
(WebCore::Quirks::shouldIgnoreViewportArgumentsToAvoidExcessiveZoom const):
(WebCore::Quirks::shouldLayOutAtMinimumWindowWidthWhenIgnoringScalingConstraints const):
* Source/WebCore/page/Quirks.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::setCanIgnoreViewportArgumentsToAvoidExcessiveZoomIfNeeded):
(WebKit::WebPage::didCommitLoad):
(WebKit::WebPage::updateWebsitePolicies):
* Tools/TestWebKitAPI/Tests/ios/Viewport.mm:
(TestWebKitAPI::makeViewportMetaTag):
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/265804@main">https://commits.webkit.org/265804@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a56cff12539a80d81ff23cc02342bcc6c4c58ffb

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/11937 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/12136 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/12513 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/13582 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/11363 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/11956 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/14525 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/12117 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/14223 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/12101 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/12914 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/10096 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/14005 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/10206 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/10832 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/17959 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/11284 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/10991 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/14159 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/11406 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/9433 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/10584 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2917 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/14866 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/11263 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -7198,7 +7198,8 @@ void WebPage::didCommitLoad(WebFrame* frame)
 
-    m_viewportConfiguration.setCanIgnoreViewportArgumentsToAvoidExcessiveZoom(shouldIgnoreMetaViewport());
+    if (auto* document = coreFrame ? coreFrame->document() : nullptr; document && document->quirks().shouldIgnoreViewportArgumentsToAvoidExcessiveZoom())
+        m_viewportConfiguration.setCanIgnoreViewportArgumentsToAvoidExcessiveZoom(shouldIgnoreMetaViewport());
```
Nit - might be good to pull this behind a static helper function to make it slightly less repetitive, e.g.:

```
static void setCanIgnoreViewportArgumentsToAvoidExcessiveZoomIfNeeded(ViewportConfiguration& configuration, Frame* frame)
{
    …
}
```

aprotyas replied:
Good idea! Implemented in [81e2290](https://github.com/WebKit/WebKit/pull/15552/commits/81e2290a4e53230b94bc0131d53dd89c8f793c43)

aprotyas committed:
Initial scale of a page is not respected when both initial scale and viewport width are set
https://bugs.webkit.org/show_bug.cgi?id=258663
rdar://111515914

Reviewed by Wenson Hsieh.

In 256476@main, we introduce a viewport sizing heuristic to clamp the
maximum initial scale of a page. This was necessitated by Wikipedia's
dynamic initial scale assignment (to maintain their viewport width
invariant of 1000) leading to Wiki articles being excessively zoomed in
on large screens (such as when an iPad is connected to an external
display).

However, the somewhat arbitrarily chosen initial scale cap of 1.2 leads
to a sub-optimal web development experience because we perform the
initial scale clamping on every page (when the `IgnoreMetaViewport` is
enabled), meaning it isn't clear why the initial scale is not being
respected, even when a developer may have had a valid use-case.

To mitigate this situation, we only apply this initial scale clamping as
a quirk for pages on the "wikipedia.org" domain. This is a relatively
safe decision given that Wikipedia is probably the only top site that
performs a dynamic initial scale assignment (that too with potentially
*large* scale values).

We also introduce an API test to assert this behavior is only limited to
Wikipedia.

* LayoutTests/fast/viewport/ios/adjust-fixed-width-and-initial-scale-to-avoid-excessive-zooming.html: Removed.

This layout test is no longer valid.

* Source/WebCore/page/Quirks.cpp:
(WebCore::isWikipediaDomain):
(WebCore::Quirks::shouldIgnoreViewportArgumentsToAvoidExcessiveZoom const):
(WebCore::Quirks::shouldLayOutAtMinimumWindowWidthWhenIgnoringScalingConstraints const):
* Source/WebCore/page/Quirks.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::setCanIgnoreViewportArgumentsToAvoidExcessiveZoomIfNeeded):
(WebKit::WebPage::didCommitLoad):
(WebKit::WebPage::updateWebsitePolicies):
* Tools/TestWebKitAPI/Tests/ios/Viewport.mm:
(TestWebKitAPI::makeViewportMetaTag):
(TestWebKitAPI::TEST):

Canonical link: https://commits.webkit.org/265804@main
